server/services/store/sqlstore/blocks.go

```diff
@@ -788,3 +802,264 @@ func (s *SQLStore) duplicateBlock(db sq.BaseRunner, boardID string, blockID stri
+		Columns(columns...).
+		Select(selectQuery)
+
+	sql, args, err := insertQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	sql, args, err = insertHistoryQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertHistoryQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	// insert into blocks table must happen before history table, otherwise the history
+	// table will be changed and the second query will fail to find the same records.
+	result, err := insertQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ := result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	result, err = insertHistoryQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ = result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertHistoryQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	return nil
+}
+
+func (s *SQLStore) findOrphansForBoards(db sq.BaseRunner, limit int) ([]string, error) {
+	if limit <= 0 {
+		limit = 1000
+	}
+	if limit > 1000 {
+		return nil, ErrLimitExceeded{1000}
+	}
+
+	// Find all orphaned child blocks for deleted boards.
+	// Squirrel can't express multiple tables in FROM, nor can it mix `From` and `FromSelect`.
+	/*
+		SELECT fb.id FROM %sblocks AS fb WHERE fb.board_id IN
+		(
+			SELECT bh.id FROM %sboards_history AS bh,
+				(SELECT id, max(insert_at) AS max_insert_at FROM %sboards_history GROUP BY id) AS sub
+			WHERE bh.id=sub.id AND bh.insert_at=sub.max_insert_at AND bh.delete_at > 0
+		)
+		ORDER BY fb.insert_at DESC LIMIT ? ;
+	*/
+	query := s.getQueryBuilder(s.db).
+		Select("fb.id").
+		From(s.tablePrefix+"blocks AS fb").
+		Suffix(fmt.Sprintf(` WHERE fb.board_id IN
+			(
+				SELECT bh.id FROM %sboards_history AS bh,
+					(SELECT id, max(insert_at) AS max_insert_at FROM %sboards_history GROUP BY id) AS sub
+				WHERE bh.id=sub.id AND bh.insert_at=sub.max_insert_at AND bh.delete_at > 0
+			)
+			ORDER BY fb.insert_at DESC LIMIT ? ;
+		`, s.tablePrefix, s.tablePrefix), limit)
+
+	rows, err := query.Query()
+	if err != nil {
+		s.logger.Error("findOrphansForBoards ERROR", mlog.Err(err))
+		return nil, err
+	}
+	defer s.CloseRows(rows)
+
+	results := []string{}
+	for rows.Next() {
+		var id string
+		err := rows.Scan(&id)
+		if err != nil {
+			s.logger.Error("findOrphansForBoards cannot scan id", mlog.Err(err))
+			return nil, err
+		}
+		results = append(results, id)
+	}
+
+	return results, nil
+}
+
+func (s *SQLStore) findOrphansForCards(db sq.BaseRunner, limit int) ([]string, error) {
+	if limit <= 0 {
+		limit = 1000
+	}
+	if limit > 1000 {
+		return nil, ErrLimitExceeded{1000}
+	}
+
+	// Find all orphaned child blocks for deleted cards.
+	// Squirrel can't express multiple tables in FROM, nor can it mix `From` and `FromSelect`.
+	/*
```

Comment:
same here about the SQL comment, it can end up outdated and leading to confusion, I prefer the textual comment on top showing the intention.